// eiyuu/build/src/class/Eiyuu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cheerio_1 = require("cheerio");
const modifier_1 = require("../modifier");
const defaults_1 = __importDefault(require("../defaults"));
const base_1 = __importDefault(require("../base"));
class Eiyuu {
    gelbooruURL;
    danbooruURL;
    hypnohubURL;
    konachanURL;
    lolibooruURL;
    rule34URL;
    realbooruURL;
    safebooruURL;
    tbibURL;
    xbooruURL;
    yandereURL;
    searchElements;
    useragent;
    searchSortings;
    searchSortingsMoebooruBased;
    followRedirects;
    /**
     * Eiyuu
     * @param useragent custom useragent
     * @param followRedirects enable HTTP redirect following
    */
    constructor(useragent, followRedirects) {
        this.useragent = useragent || defaults_1.default.useragent;
        this.followRedirects = followRedirects || defaults_1.default.followRedirects;
        this.searchElements = "table.highlightable";
        this.searchSortings = "*&sort=desc&order_by=index_count";
        this.searchSortingsMoebooruBased = "*&type=&order=count";
        this.gelbooruURL = base_1.default.GELBOORU;
        this.danbooruURL = base_1.default.DANBOORU;
        this.hypnohubURL = base_1.default.HYPNOHUB;
        this.konachanURL = base_1.default.KONACHAN;
        this.lolibooruURL = base_1.default.LOLIBOORU;
        this.rule34URL = base_1.default.RULE34;
        this.realbooruURL = base_1.default.REALBOORU;
        this.safebooruURL = base_1.default.SAFEBOORU;
        this.tbibURL = base_1.default.TBIB;
        this.xbooruURL = base_1.default.XBOORU;
        this.yandereURL = base_1.default.YANDERE;
    }
    /**
     * Search arbitrary query on danbooru.
     * @param query The query to search.
     * @example
     * ```js
     * search.danbooru("amber").then(res => { console.log(res); });
     * ```
     * https://danbooru.donmai.us/tags?commit=Search&search%5Bhide_empty%5D=yes&search%5Bname_or_alias_matches%5D=
     */
    async danbooru(query) {
        try {
            const res = await (0, modifier_1.request)(this.danbooruURL, `*${query}`, "*%2A&search%5Border%5D=count", this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("/posts?tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on gelbooru.
     * @param query The query to search.
     * @example
     * ```js
     * search.gelbooru("amber").then(res => { console.log(res); });
     * ```
     * https://gelbooru.com/index.php?page=tags&s=list&tags=
     */
    async gelbooru(query) {
        try {
            const res = await (0, modifier_1.request)(this.gelbooruURL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on hypnohub.
     * @param query The query to search.
     * @example
     * ```js
     * search.hypnohub("erza").then(res => { console.log(res); });
     * ```
     * https://hypnohub.net/index.php?page=tags&s=list&tags=
     */
    async hypnohub(query) {
        try {
            const res = await (0, modifier_1.request)(this.hypnohubURL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on konachan.
     * @param query The query to search.
     * @example
     * ```js
     * search.konachan("amber").then(res => { console.log(res); });
     * ```
     * https://konachan.com/tag?name=
     */
    async konachan(query) {
        try {
            const res = await (0, modifier_1.request)(this.konachanURL, `*${query}`, this.searchSortingsMoebooruBased, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $("tbody").find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("/post?tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on lolibooru.
     * @param query The query to search.
     * @example
     * ```js
     * search.lolibooru("kanna").then(res => { console.log(res); });
     * ```
     * https://lolibooru.moe/tag?name=
     */
    async lolibooru(query) {
        try {
            const res = await (0, modifier_1.request)(this.lolibooruURL, `*${query}`, this.searchSortingsMoebooruBased, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $("tbody").find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("/post?tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on rule34.
     * @param query The query to search.
     * @example
     * ```js
     * search.rule34("lisa").then(res => { console.log(res); });
     * ```
     * https://rule34.xxx/index.php?page=tags&s=list&tags=
     */
    async rule34(query) {
        try {
            const res = await (0, modifier_1.request)(this.rule34URL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on realbooru.
     * @param query The query to search.
     * @example
     * ```js
     * search.realbooru("anri").then(res => { console.log(res); });
     * ```
     * https://realbooru.com/index.php?page=tags&s=list&tags=
     */
    async realbooru(query) {
        try {
            const res = await (0, modifier_1.request)(this.realbooruURL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on safebooru.
     * @param query The query to search.
     * @example
     * ```js
     * search.safebooru("scathach").then(res => { console.log(res); });
     * ```
     * https://safebooru.org/index.php?page=tags&s=list&tags=
     */
    async safebooru(query) {
        try {
            const res = await (0, modifier_1.request)(this.safebooruURL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on tbib.
     * @param query The query to search.
     * @example
     * ```js
     * search.tbib("erza").then(res => { console.log(res); });
     * ```
     * https://tbib.org/index.php?page=tags&s=list&tags=
     */
    async tbib(query) {
        try {
            const res = await (0, modifier_1.request)(this.tbibURL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on xbooru.
     * @param query The query to search.
     * @example
     * ```js
     * search.xbooru("amber").then(res => { console.log(res); });
     * ```
     * https://xbooru.com/index.php?page=tags&s=list&tags=
     */
    async xbooru(query) {
        try {
            const res = await (0, modifier_1.request)(this.xbooruURL, `*${query}`, this.searchSortings, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $(this.searchElements).find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("&tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
    /**
     * Search arbitrary query on yandere.
     * @param query The query to search.
     * @example
     * ```js
     * search.yandere("eula").then(res => { console.log(res); });
     * ```
     * https://yande.re/tag?name=
     */
    async yandere(query) {
        try {
            const res = await (0, modifier_1.request)(this.yandereURL, `*${query}`, this.searchSortingsMoebooruBased, this.useragent, this.followRedirects);
            const $ = (0, cheerio_1.load)(res.body);
            const gets = $("tbody").find("a").map((i, el) => $(el).attr("href")).get();
            const tags = gets.map(el => el.split("/post?tags=")[1]).filter(el => el !== undefined);
            const replaced = tags.map(el => (0, modifier_1.htmlDecode)(el));
            return replaced;
        }
        catch (e) {
            const error = e;
            throw new Error(error);
        }
    }
}
exports.default = Eiyuu;
